tests/test_factorgraph.py

```diff
@@ -904,3 +904,103 @@ def test_and_rounding_error_simple():
     bp.bp_loopy(5, initialize_states=False)
     assert (bp.get_distribution("K0") >= 0.0).all()
     assert (bp.get_distribution("K1") >= 0.0).all()
+
+
+def test_dampening_nochange():
+    graph = """
+        NC 2
+        PROPERTY s1: x = a^b
+        VAR MULTI x
+        VAR MULTI a
+        VAR MULTI b
+        """
+    distri_a = np.array([[0.6, 0.4]])
+    distri_b = np.array([[0.8, 0.2]])
+    graph = FactorGraph(graph)
+    n = 1
+    bp_state_no_dampen = BPState(graph, n)
+    bp_state_no_dampen.set_evidence("a", distri_a)
+    bp_state_no_dampen.set_evidence("b", distri_b)
+    bp_state_full_dampen = BPState(graph, n)
+    bp_state_full_dampen.set_evidence("a", distri_a)
+    bp_state_full_dampen.set_evidence("b", distri_b)
+
+    bp_state_no_dampen.bp_loopy(5, initialize_states=False)
+    bp_state_full_dampen.bp_loopy(5, initialize_states=False, alpha=1.0)
+    assert np.allclose(
+        bp_state_no_dampen.get_distribution("x"),
+        bp_state_full_dampen.get_distribution("x"),
+    )
+
+
+def test_clear_beliefs():
+    graph = """
+    NC 2
+    PROPERTY s1: x = a^b
+    VAR MULTI x
+    VAR MULTI a
+    VAR MULTI b
+    """
+    fg = FactorGraph(graph)
+    bp = BPState(fg, 1)
+    for k in fg.vars():
+        bp.set_evidence(k, make_distri(2, 1))
+    bp.bp_loopy(1, False, clear_beliefs=False)
+
+    assert bp.get_belief_from_var("x", "s1") is not None
+    assert bp.get_belief_from_var("a", "s1") is not None
+    assert bp.get_belief_from_var("b", "s1") is not None
+
+
+def test_dampening_correctness():
+    factor_graph = """NC 16
+    VAR MULTI K0
+    VAR MULTI K1
+    VAR MULTI L1
+    VAR MULTI N1
+    VAR MULTI B
+    VAR MULTI C
+    VAR MULTI N0
+    VAR MULTI A
+    VAR MULTI L2
+    VAR MULTI L3
+    PUB SINGLE IV
+    PROPERTY P1: L1 = K0 ^ K1
+    PROPERTY P2: N1 = !K1
+    PROPERTY P3: B = N1 & IV
+    PROPERTY P4: C = B ^ K0
+    PROPERTY P5: N0 = !K0
+    PROPERTY P6: A = N0 & K1
+    PROPERTY P7: L2 = A ^ C
+    PROPERTY P8: L3 = K1 ^ C"""
+
+    fg = FactorGraph(factor_graph)
+    alpha = 0.9
+    bp_dampen = BPState(fg, 1, public_values={"IV": 0xC})
+    bp_no_dampen = BPState(fg, 1, public_values={"IV": 0xC})
+    for k in fg.vars():
+        d = make_distri(16, 1)
+        bp_no_dampen.set_evidence(k, distribution=d)
+        bp_dampen.set_evidence(k, distribution=d)
+
+    for v in fg.vars():
+        bp_dampen.propagate_var(v)
+        bp_no_dampen.propagate_var(v)
+    prev = bp_no_dampen.get_belief_from_var("A", "P7")
+    prev2 = bp_no_dampen.get_belief_from_var("C", "P4")
+    prev3 = bp_no_dampen.get_belief_from_var("K0", "P1")
+    bp_dampen.bp_loopy(1, False, alpha=alpha, clear_beliefs=False)
+    bp_no_dampen.bp_loopy(1, False, clear_beliefs=False)
+
+    assert np.allclose(
```

Comment:
For the test comprehensiveness, it would be good if these were done on all edges of the graph.

Also, this does not test the case where there are non-uniform beliefs in `factors -> vars`. Therefore, it would be good to add a second part of the test which does the same thing, but after one full BP iteration.